### ISSUE TITLE
Use something like vscodeignore instead of npmignore for ignored files?

### ISSUE THREAD

**Issue by DanTup** [microsoft/vscode-dev-containers#45]
In https://github.com/microsoft/vscode-dev-containers/blob/master/CONTRIBUTING.md it says:

> VS Code respects .npmignore as a way to keep certain content out of projects when your definition is used.

It feels really weird to put npmignore files in projects that have nothing to do with NPM and VS Code to use them. VS Code already has its own ignore file (for publishing extensions) so it seems like using that (or something else project-agnostic) would make much more sense than this? When tweaking the Dart container, I originally deleted this file thinking it was accidentally left over from starting from a node container!

(Also, the docs should be more clear exactly what this is doing, it's not really clear to me from that description exactly when those files are being ignored ignored?).

**Comment by Chuxel** [comment]
@DanTup Yeah, currently the contents of `.npmignore` are used in pulling the definition into the UX in the Remote - Containers extension when you run the Remote-Conatiners: Add Container Configuration File... or open / reopen a folder in container when there's no definition in the folder.  There is a npm pacakge that gets generated to cause this to happen. We can definitely make that clearer in CONTRIBUTING.

The name of the file is a good point. We've gone back and forth on this one a bit -- your insight here is great since it gives us a sense of first reactions. We'd been thinking about a unique filename (e.g. .devconignore or .remotedevignore).  @chrmarti We could just have a pre-package step that renames this alternate filename to .npmignore before packaging since it sounds like that may be less confusing.  What do you think?

**Comment by DanTup** [comment]
> There is a npm pacakge that gets generated to cause this to happen

Ah, that makes more sense now then :-) I still think it's weird though, so I think it's better to hide that implementation detail.

Just to confirm I understand this correctly - the folders in this repo (for ex `dart`) are serving both as fully-contained samples (with the test-project folder), but also can be used with arbitrary folders (in which case the definition is used, but the test-project folder is not), and the npmignore is to exclude the "sample project" files from the workspace when it's used to open another folder, right?

(Also - slight tangent - what's the relationship between this repo and the all the `-try` repos?)

**Comment by Chuxel** [comment]
>Just to confirm I understand this correctly - the folders in this repo (for ex dart) are serving both as fully-contained samples (with the test-project folder), but also can be used with arbitrary folders (in which case the definition is used, but the test-project folder is not), and the npmignore is to exclude the "sample project" files from the workspace when it's used to open another folder, right?

@DanTup Yeah its primary purpose is really to provide general purpose configurations for people to use in their own projects.  This is why the project assets are not in the root of the folder. The test project folder is primarily there for contributors and reviewers to test the configurations. They can also be used by customers to see one of the definitions in action, but not all definitions have them and it's not strictly required.

>(Also - slight tangent - what's the relationship between this repo and the all the -try repos?)

The `-try` repos are for full sample projects that use a dev container.  Their goal is that you can just "clone and go" to try out the capabilities of the extension.  This repository is for general purpose definitions designed to be used in the developer's own project.  This is why they do not expose ports and other things referenced in the test project.  The directions for using them is to use the built in VS Code Remote - Containers command or to just copy the contents of the `.devcontainer` folder directly into their existing project folder.

At one point in this repo's history, it served both purposes. However, over time we realized that the contents of the definitions needed to be fundamentally different for a sample verses a "drop in" starting point for an existing project.  In some cases the code is similar, but the `devcontainer.json` (and in some cases other files in the repo) is typically different for these reasons.

**Comment by DanTup** [comment]
@Chuxel thanks, that makes more sense! I got a bit confused when testing the Dart one about how the lifetime of the container and the open folder were related (I thought they were the same, but when I closed a Folder, I was still connected to the container!), but it makes more sense now I realise they're separate, but just bundled together here for convenience :-)

**Comment by chrmarti** [comment]
I'd keep `.npmignore` while we are using NPM, that makes it obvious where to look for documentation on the supported syntax.

**Comment by DanTup** [comment]
> that makes it obvious where to look for documentation on the supported syntax.

That's a fair point I hadn't considered. Maybe adding a comment to the top of the npmignore file that explains what it's used for then is all that's needed (although it's in the docs, it's easy to just take a copy of the folder and assume it's not needed for your non-NPM project).
